yt_dlp/downloader/common.py

```diff
@@ -72,12 +72,30 @@ class FileDownloader:
 
     def __init__(self, ydl, params):
         """Create a FileDownloader object with the given options."""
-        self.ydl = ydl
+        self._set_ydl(ydl)
         self._progress_hooks = []
         self.params = params
         self._prepare_multiline_status()
         self.add_progress_hook(self.report_progress)
 
+    def _set_ydl(self, ydl):
+        self.ydl = ydl
+
+        for func in (
+            'deprecation_warning',
+            'report_error',
+            'report_file_already_downloaded',
+            'report_warning',
+            'to_console_title',
+            'to_stderr',
+            'trouble',
+            'write_debug',
+        ):
+            setattr(self, func, getattr(ydl, func))
```

Comment:
I think the best way to do logging would be to split off the logging functionality into a separate object that only deals with logging and nothing else, and just pass that object around to whoever needs it. And then do away with the mirroring. It would be very easy to adapt `yt_dlp.cookies` to this style at least.

Similar strategy could be applied to network requests; I believe @coletdjnz’s draft PRs follow that route. We do this enough and we’ll discover that `YoutubeDL` has basically no responsibilities other than being a grab-bag of other objects…

Reply:
> Similar strategy could be applied to network requests; I believe @coletdjnz’s draft PRs follow that route.

Yeah... I literally just copied in these 'mirroring' functions from FD... I agree it would be better to pass around some sort of logging object (or some global one that is imported?). 

There is the official python logging library; is there any reason we don't use that with our own custom logger?

Reply:
I’m not sure the built-in `logging` would necessarily fit yt-dlp very well.

yt-dlp two kinds of messages: event messages and progress messages. Event messages are one-shot and permanent; progress messages are ephemeral and re-written to reflect current state. `logging` seems to be built to only handle event messages.

If we get parallel downloading some day, we would probably like for each downloader to display its own progress bar, while allowing potential intervening log messages to show as well without corrupting the display. I don’t think `logging` was built to handle such situations.